### ISSUE TITLE
TodoLite example does not work on AWS server

### ISSUE THREAD

**Issue by FedericoAguirre** [couchbase/couchbase-lite-java-core#665]
We  ran ToDoLite example for Android on two tablets and the example worked well pointing out to URL http://demo.couchbasemobile.com:4984/todolite/, doing push and pull replication, but when we tried to set up the url pointing out to our server, the demo just did the push replication; the pull replication was not possible.
We got the next error:

00:01:32.750442 Sync fn rejected: new=map[_rev:4-7ff3be0eb78c893d168bad6172f744f1 _id:p:10153982788344535 _revisions:map[start:4 ids:[7ff3be0eb78c893d168bad6172f744f1 24bf9af1e209120a955d68892e7011a0 961bcfce612805ccad7e8f500bae7a12 2eb71185535395410249c5646fe8fcd2]] name:Javier Linares Villuendas user_id:10153982788344535 device_tokens:[APA91bG791xAdy_qatfciKjxK831_GOKJEyjgv5FRi18V3TLCFafiJT_rOAuD0PNvT3hI7K1hlLJXeEANC0-Z7BM2HnOU_j8xQz3IwZPbhkKxiFAuQM-gsA7hduU9JHHIvWbU71XSXxo APA91bG791xAdy_qatfciKjxK831_GOKJEyjgv5FRi18V3TLCFafiJT_rOAuD0PNvT3hI7K1hlLJXeEANC0-Z7BM2HnOU_j8xQz3IwZPbhkKxiFAuQM-gsA7hduU9JHHIvWbU71XSXxo APA91bG791xAdy_qatfciKjxK831_GOKJEyjgv5FRi18V3TLCFafiJT_rOAuD0PNvT3hI7K1hlLJXeEANC0-Z7BM2HnOU_j8xQz3IwZPbhkKxiFAuQM-gsA7hduU9JHHIvWbU71XSXxo] type:profile]  old= --> 403 wrong user
00:01:32.750470         BulkDocs: Doc "p:10153982788344535" --> 403 wrong user (403 wrong user)

We used an Amazon AMI with Couchbase Server and Sync gateway version 3.
We used the next config file:

{
    "log":["REST", "Shadow", "CRUD", "CRUD+", "HTTP", "HTTP+", "Access", "Changes", "Changes+","Auth", "SSL"],
    "facebook" : {
        "register" : true
    },
	"databases": {
    "todolite": {
      "server":"http://ecX-XX-X-XXX-XX.compute-1.amazonaws.com:8091",
      "bucket":"todolite",
      "users": {
        "GUEST": {"disabled": true}
      },
      "sync": 
  	`
      function(doc, oldDoc) {
        // NOTE this function is the same across the iOS, Android, and PhoneGap versions.
        if (doc.type == "task") {
          if (!doc.list_id) {
            throw({forbidden : "Items must have a list_id"})
          }
          channel("list-"+doc.list_id);
        } else if (doc.type == "list") {
          channel("list-"+doc._id);
          if (!doc.owner) {
            throw({forbidden : "List must have an owner"})
          }
          if (oldDoc) {
            var oldOwnerName = oldDoc.owner.substring(oldDoc.owner.indexOf(":")+1);
            requireUser(oldOwnerName)
          }
          var ownerName = doc.owner.substring(doc.owner.indexOf(":")+1);
          access(ownerName, "list-"+doc._id);
          if (Array.isArray(doc.members)) {
            var memberNames = [];
            for (var i = doc.members.length - 1; i >= 0; i--) {
              memberNames.push(doc.members[i].substring(doc.members[i].indexOf(":")+1))
            };
            access(memberNames, "list-"+doc._id);
          }
        } else if (doc.type == "profile") {
          channel("profiles");
          var user = doc._id.substring(doc._id.indexOf(":")+1);

          if (user !== doc.user_id) {
            throw({forbidden : "Profile user_id must match docid : " + user + " : " + doc.user_id})
          }
          requireUser(user);
          access(user, "profiles"); // TODO this should use roles
        }
      }
    `
    }
  }
}

The same program did not work as expected.

The config file was taken from https://github.com/couchbase/sync_gateway/blob/master/examples/democlusterconfig.json it was just adapted to our needs.

Can you help us?

**Comment by hideki** [comment]
Hi @FedericoAguirre,

ToDoLite sample app uses Facebook login. It might need additional setting for Facebook login.

@tleyden Could you please help this user?

Thanks!

**Comment by jamiltz** [comment]
The 403 wrong user message message was thrown by the call to `requireUser(user)` , so perhaps, the user with name `10153982788344535 ` doesn't exist in the database. You can make a request on the admin port of SG to get the list of users `curl http://localhost:4985/db/_user/`. Is this user name in the array returned in the response?

**Comment by jalinvi** [comment]
Hi all!

I'm @FedericoAguirre  coworker. I would like to give more context of the scenario describe above by Fede:

1. We downloaded TodoLite source code for Android and we installed in two diferentes devices pointing out to URL http://demo.couchbasemobile.com:4984/todolite/
2. We logged in using the same facebook account at the same time in both devices. The replication worked almost in real time. When you added new lists or tasks in one device you could see instantly the changes in the other devices. Everything was working fine.
3. We changed the URL, pointing to our Amazon Couchbase Server with the config as described by @FedericoAguirre.
4. We logged in again using via facebook in both devices. In this case, if I added new lists or task, they didn't go from one device to the other. The worked as if they where different accounts. When I logged out/in each devices just downloaded the lists/tasks they previously created.

We are assuming we have some missing or wrong server set up. @jamiltz @tleyden could you please assist us? any idea what we're doing wrong?

Thank so much!

**Comment by FedericoAguirre** [comment]
Hello @jamiltz.

About your question when I ask about the user with curl I get:

14:42:06.117703 HTTP:  #066: GET /todolite/:10205046437374692  (ADMIN)
14:42:06.118076 HTTP: #066:     --> 404 missing  (0.4 ms)
__{"error":"not_found","reason":"missing"}__

The config file contains the next JSON:

{
    "facebook" : {
        "register" : true
    },
"log":["*"],
        "databases": {
    "todolite": {
      "server":"http://ec2-XX-X-XXX-XX.compute-1.amazonaws.com:8091",
      "bucket":"todolite",
      "users": {
        "GUEST": {"disabled": true}
      },
      "sync":
        `
      function(doc, oldDoc) {
        // NOTE this function is the same across the iOS, Android, and PhoneGap versions.
        if (doc.type == "task") {
          if (!doc.list_id) {
            throw({forbidden : "Items must have a list_id"})
          }
          channel("list-"+doc.list_id);
        } else if (doc.type == "list") {
          channel("list-"+doc._id);
          if (!doc.owner) {
            throw({forbidden : "List must have an owner"})
          }
          if (oldDoc) {
            var oldOwnerName = oldDoc.owner.substring(oldDoc.owner.indexOf(":")+1);
            requireUser(oldOwnerName)
          }
          var ownerName = doc.owner.substring(doc.owner.indexOf(":")+1);
          access(ownerName, "list-"+doc._id);
          if (Array.isArray(doc.members)) {
            var memberNames = [];
            for (var i = doc.members.length - 1; i >= 0; i--) {
              memberNames.push(doc.members[i].substring(doc.members[i].indexOf(":")+1))
            };
            access(memberNames, "list-"+doc._id);
          }
        } else if (doc.type == "profile") {
          channel("profiles");
          var user = doc._id.substring(doc._id.indexOf(":")+1);

          if (user !== doc.user_id) {
            throw({forbidden : "Profile user_id must match docid : " + user + " : " + doc.user_id})
          }
          requireUser(user);
          access(user, "profiles"); // TODO this should use roles
        }
      }
    `
    }
  }
}

**Comment by FedericoAguirre** [comment]
Hello @hideki, @jamiltz @tleyden.

We still having the same trouble, and the user does not appears while calling by curl.

PS. I closed the issue by mistake.

**Comment by tleyden** [comment]
> We used an Amazon AMI with Couchbase Server and Sync gateway version 3.

Unfortunately that Sync Gateway version (1.0.3), is out of date and probably isn't compatible with the current couchbase lite codebase.

We are working on rebuilding a newer AMI -- in the meantime, can you download the latest Sync Gateway release and replace the older Sync Gateway 1.0.3?  We can provide help if needed.

/cc @zgramana

**Comment by FedericoAguirre** [comment]
@tleyden, which version of Couchbase server and sync gateway do you recommend to use? Can you provide the proper urls to download?

**Comment by zgramana** [comment]
@federicaaguirre latest was just released on http://couchbase.com/downloads

**Comment by FedericoAguirre** [comment]
We solved this issue after different trials. The final good result was: an AMI machine with Ubuntu 14.0.4, Couchbase server 3.0.3 and Sync gateway 1.1.0.
We found that some characters in config files were not parsed correctly by sync gateway like \r or " when copying files from examples folder. We replaced those characters in config file and saved with UTF8 encoding and the TodoLite app worked ok.
